# Hail-Powered Science

The following is an incomplete list of scientific work enabled by Hail. We welcome you to add additional examples by [editing this page directly](https://github.com/hail-is/hail/edit/master/hail/www/references.md), after which we will review the pull request to confirm the addition is valid. Please adhere to the existing formatting conventions.

In addition to software development, the Hail team engages in theoretical, algorithmic, and empirical research inspired by scientific collaboration. Examples include [Loss landscapes of regularized linear autoencoders](https://github.com/danielkunin/Regularized-Linear-Autoencoders), [Secure multi-party linear regression at plaintext speed](https://github.com/jbloom22/DASH), and [A synthetic-diploid benchmark for accurate variant-calling evaluation](https://www.nature.com/articles/s41592-018-0054-7).

## 2019

* Huang, Qin, et al. "Delivering genes across the blood-brain barrier: LY6A, a novel cellular receptor for AAV-PHP. B capsids." *bioRxiv* (2019): 538421. https://www.biorxiv.org/content/10.1101/538421v1

* Kurki, Mitja I., et al. "Contribution of rare and common variants to intellectual disability in a sub-isolate of Northern Finland." *Nature Communications* 10.1 (2019): 410. https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/30679432/

* Martin, Alicia R., et al. "Current clinical use of polygenic scores will risk exacerbating health disparities." *bioRxiv* (2019): 441261. https://www.biorxiv.org/content/10.1101/441261v3

* Collaborative, Epi25, et al. "Ultra-rare genetic variation in the epilepsies: a whole-exome sequencing study of 17,606 individuals." *American Journal of Human Genetics* (2019): https://www.cell.com/ajhg/fulltext/S0002-9297(19)30207-1

* Karczewski, Konrad J., et al. "Variation across 141,456 human exomes and genomes reveals the spectrum of loss-of-function intolerance across human protein-coding genes." *bioRxiv* (2019): 531210. https://www.biorxiv.org/content/10.1101/531210v2.abstract

* Whiffin, Nicola, et al. "Human loss-of-function variants suggest that partial LRRK2 inhibition is a safe therapeutic strategy for Parkinsons disease." *bioRxiv*() (2019): 561472. https://www.biorxiv.org/content/10.1101/561472v1

* Cummings, Beryl B., et al. "Transcript expression-aware annotation improves rare variant discovery and interpretation." *bioRxiv* (2019): 554444. https://www.biorxiv.org/content/10.1101/554444v1

* Wang, Qingbo, et al. "Landscape of multi-nucleotide variants in 125,748 human exomes and 15,708 genomes." *bioRxiv* (2019): 573378. https://www.biorxiv.org/content/10.1101/573378v2

* Minikel, Eric Vallabh, et al. "Evaluating potential drug targets through human loss-of-function genetic variation." *bioRxiv* (2019): 530881. https://www.biorxiv.org/content/10.1101/530881v2

* Collins, Ryan L., et al. "An open resource of structural variation for medical and population genetics." *bioRxiv* (2019): 578674. https://www.biorxiv.org/content/10.1101/578674v1

* Whiffin, Nicola, et al. "Characterising the loss-of-function impact of 5'untranslated region variants in whole genome sequence data from 15,708 individuals." *bioRxiv* (2019): 543504. https://www.biorxiv.org/content/10.1101/543504v1

* Lacaze, Paul, et al. "The Medical Genome Reference Bank: a whole-genome data resource of 4000 healthy elderly individuals. Rationale and cohort design." *European Journal of Human Genetics* 27.2 (2019): 308. https://www.nature.com/articles/s41431-018-0279-z

* Cirulli, Elizabeth T., et al. "Genome-wide rare variant analysis for thousands of phenotypes in 54,000 exomes." *bioRxiv* (2019): 692368. https://www.biorxiv.org/content/10.1101/692368v1.abstract

* Kerminen, Sini, et al. "Geographic Variation and Bias in the Polygenic Scores of Complex Diseases and Traits in Finland." *American Journal of Human Genetics* (2019). https://www.biorxiv.org/content/10.1101/485441v1.abstract

* Jiang, Fan, Kyle Ferriter, and Claris Castillo. "PIVOT: Cost-Aware Scheduling of Data-Intensive Applications in a Cloud-Agnostic System." https://renci.org/wp-content/uploads/2019/02/Cloud_19.pdf

* Pividori, Milton, et al. "Shared and distinct genetic risk factors for childhood-onset and adult-onset asthma: genome-wide and transcriptome-wide studies." *Lancet Respiratory Medicine* 7.6 (2019): 509-522. https://www.biorxiv.org/content/10.1101/427427v2

* Cox, Samantha L., et al. "Genetic contributions to variation in human stature in prehistoric Europe." *bioRxiv* (2019): 690545. https://www.biorxiv.org/content/10.1101/690545v1.abstract

* Abrar, Faheem. A Modular Parallel Pipeline Architecture for GWAS Applications in a Cluster Environment. Diss. University of Saskatchewan, 2019. https://harvest.usask.ca/handle/10388/12087

* Khera, Amit V., et al. "Whole-genome sequencing to characterize monogenic and polygenic contributions in patients hospitalized with early-onset myocardial infarction." *Circulation* 139.13 (2019): 1593-1602. https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.118.035658

## 2018

* Molnos, Sophie Claudia. Metabolites: implications in type 2 diabetes and the effect of epigenome-wide interaction with genetic variation. Diss. Technische Universität München, 2018. https://mediatum.ub.tum.de/1372795f

* Bis, Joshua C., et al. "Whole exome sequencing study identifies novel rare and common Alzheimer’s-associated variants involved in immune response and transcriptional regulation." *Molecular Psychiatry* (2018): 1. https://www.nature.com/articles/s41380-018-0112-7

* Gormley, Padhraig, et al. "Common variant burden contributes to the familial aggregation of migraine in 1,589 families." *Neuron* 98.4 (2018): 743-753. https://www.ncbi.nlm.nih.gov/pubmed/30189203

* Rivas, Manuel A., et al. "Insights into the genetic epidemiology of Crohn's and rare diseases in the Ashkenazi Jewish population." *PLoS Genetics* 14.5 (2018): e1007329. https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007329

* Satterstrom, F. Kyle, et al. "ASD and ADHD have a similar burden of rare protein-truncating variants." *bioRxiv* (2018): 277707. https://www.biorxiv.org/content/10.1101/277707v1

* Zekavat, Seyedeh M., et al. "Deep coverage whole genome sequences and plasma lipoprotein (a) in individuals of European and African ancestries." *Nature Communications* 9.1 (2018): 2606. https://www.nature.com/articles/s41467-018-04668-w

* Natarajan, Pradeep, et al. "Deep-coverage whole genome sequences and blood lipids among 16,324 individuals." *Nature Communications* 9.1 (2018): 3391. https://www.nature.com/articles/s41467-018-04668-w

* Farhan, Sali MK, et al. "Enrichment of rare protein truncating variants in amyotrophic lateral sclerosis patients." *bioRxiv* (2018): 307835. https://www.biorxiv.org/content/10.1101/307835v2

* Ganna, Andrea, et al. "Quantifying the impact of rare and ultra-rare coding variation across the phenotypic spectrum." *American Journal of Human Genetics* 102.6 (2018): 1204-1211. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5992130/

* Khera, Amit V., et al. "Genome-wide polygenic scores for common diseases identify individuals with risk equivalent to monogenic mutations." *Nature Genetics* 50.9 (2018): 1219. https://www.nature.com/articles/s41588-018-0183-z?_ga=2.263293700.980063710.1543017600-1151073636.1543017600

* Roselli, Carolina, et al. "Multi-ethnic genome-wide association study for atrial fibrillation." *Nature Genetics* 50.9 (2018): 1225. https://www.nature.com/articles/s41588-018-0133-9

* Arachchi, Harindra, et al. "matchbox: An open‐source tool for patient matching via the Matchmaker Exchange." *Human Mutation* 39.12 (2018): 1827-1834. https://onlinelibrary.wiley.com/doi/abs/10.1002/humu.23655

* Laisk, Triin, et al. "GWAS meta-analysis highlights the hypothalamic-pituitary-gonadal axis (HPG axis) in the genetic regulation of menstrual cycle length." *bioRxiv* (2018): 333708. https://www.biorxiv.org/content/10.1101/333708v1.abstract

* Rees, Elliott, et al. "Association between schizophrenia and both loss of function and missense mutations in paralog conserved sites of voltage-gated sodium channels." *bioRxiv* (2018): 246850. https://www.biorxiv.org/content/10.1101/246850v1.abstract

* Haas, Mary E., et al. "Genetic association of albuminuria with cardiometabolic disease and blood pressure." *American Journal of Human Genetics* 103.4 (2018): 461-473. https://www.cell.com/ajhg/pdf/S0002-9297(18)30270-2.pdf

* Abel, Haley J., et al. "Mapping and characterization of structural variation in 17,795 deeply sequenced human genomes." *bioRxiv* (2018): 508515. https://www.biorxiv.org/content/10.1101/508515v1.abstract

* Lane, Jacqueline M., et al. "Biological and clinical insights from genetics of insomnia symptoms." *bioRxiv* (2018): 257956. https://www.biorxiv.org/content/10.1101/257956v1.abstract

* Pividori, Milton, et al. "Shared and distinct genetic risk factors for childhood onset and adult onset asthma." *bioRxiv* (2018): 427427. https://www.biorxiv.org/content/10.1101/427427v1.abstract

## 2017

* Lessard, Samuel, et al. "Human genetic variation alters CRISPR-Cas9 on-and off-targeting specificity at therapeutically implicated loci." *Proceedings of the National Academy of Sciences* 114.52 (2017): E11257-E11266. https://www.pnas.org/content/114/52/E11257.long

## 2016

* Ganna, Andrea, et al. "Ultra-rare disruptive and damaging mutations influence educational attainment in the general population." *Nature Neuroscience* 19.12 (2016): 1563. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5127781/
